### PR TITLE
ROX-20757: Enable multiarch for basic-component-pipeline

### PR DESCRIPTION
**Pull request by Molter73**
### Description

Enable multiarch builds for the basic-component-pipeline, which controls builds for:
- scanner-v4-db
- central-db
- roxctl
- stackrox-operator

The new workflow consists of building images for amd64 in the same way that was being done before these changes and using the buildah-remote bundle for ppc64le, s390x and arm64. Once built, a manifest is created for grouping all 4 images. On top of this, the needed `konflux-` tags area applied to all archs and a separate manifest is created for these.

### User-facing documentation

(*must be* 2 items and both *must be* checked)

- [x] CHANGELOG update is not needed
- [x] Documentation is not needed

### Testing

- [x] inspected CI results

#### Automated testing

(*must be* at least 1 item and all items *must be* checked)

- [x] contributed **no automated tests**
  
This PR only changes konflux CI, no needed tests should be needed other than a successful build.

#### How I validated my change

Ran the new workflows and checked quay to see all images and manifest have been successfully published.

### CODE REVIEW

File: .tekton/basic-component-pipeline.yaml
Comment (Stringy) [reply .tekton/basic-component-pipeline.yaml:104]: Same for the post-build tasks as well

**Review by Stringy — APPROVED** [review]
LGTM! just one question around the amd64 results use. I'll defer to @stackrox/rhtap-maintainers for the definitive review

File: .tekton/basic-component-pipeline.yaml
Comment (Molter73) [reply .tekton/basic-component-pipeline.yaml:104]: We don't have JAVA dependencies, so it doesn't apply to this particular line. The prefetch of dependencies does copy the go dependencies, but those are fine because the same code is used for all archs AFAIK.

The post-build tasks were directly copied from the examples we were given so 🤷🏻‍♂️ , I need the opinion of experts on this one.

File: .tekton/basic-component-pipeline.yaml
Comment (msugakov) [inline .tekton/basic-component-pipeline.yaml:445]: I have the same question as in the PR of @Stringy 
https://github.com/stackrox/scanner/pull/1574#discussion_r1688386041

File: .tekton/basic-component-pipeline.yaml
Comment (msugakov) [reply .tekton/basic-component-pipeline.yaml:108]: Yes, we don't have Java dependencies so following the example like you did should be fine.

For post-tasks, I suppose, the ones that deal with container scanning seem like need one container, not all. There's no conceptual difference between giving them `amd64` container and giving them manifest because `amd64` one will be selected from the manifest according to the arch of the node where the code executes.

I don't know if these tasks are actually able to work with manifests or there's some limitation that they need only actual images. Using manifests seems cleaner so I wouldn't be at all opposed if you try (but you don't have to).

Comment (msugakov) [reply .tekton/basic-component-pipeline.yaml:108]: Oh, I see that you actually did.

File: .tekton/basic-component-pipeline.yaml
Comment (msugakov) [inline .tekton/basic-component-pipeline.yaml:212]: In the mean time, operator-bundle builds arrived and they bring `output-image-tag-construction` thingy. See https://github.com/stackrox/stackrox/pull/12037/files#diff-852bc59fb24fb28fe17c4ecfb8332be7dbe6476cc8a52c888cba481b2c2aeb99R198

You should use `$(params.output-image-repo):$(tasks.output-image-tag-construction.results.FINAL_IMAGE_TAG)-amd64` here and in similar places for arch-specific images, not `tasks.determine-image-tag.results.IMAGE_TAG`.

----

TIL that operator-bundle being `FROM scratch`-based container with only data still comes for each arch and with a manifest.

```json
$ podman manifest inspect registry.redhat.io/advanced-cluster-security/rhacs-operator-bundle:4.5.0    
{
    "schemaVersion": 2,
    "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
    "manifests": [
        {
            "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
            "size": 426,
            "digest": "sha256:afe3518921c5f14f11f7543f384c6cfc32eaae86d2a38bb6e36b783aeec53f79",
            "platform": {
                "architecture": "amd64",
                "os": "linux"
            }
        },
        {
            "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
            "size": 426,
            "digest": "sha256:b235dddbd9b87d565ccd2366ad615a825736c9f8126da62dfec1895c032a5a37",
            "platform": {
                "architecture": "ppc64le",
                "os": "linux"
            }
        },
        {
            "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
            "size": 426,
            "digest": "sha256:465631cade0b59189d12075ece6c4e1e18653fa61c99fa1c5c4bf8adaecc832b",
            "platform": {
                "architecture": "s390x",
                "os": "linux"
            }
        }
    ]
}
```

File: .tekton/basic-component-pipeline.yaml
Comment (msugakov) [inline .tekton/basic-component-pipeline.yaml:278]: [optional]
Looks like there's more fresh version of the task already although it does not change anything significant for multi-arch builds. Only **if you have extra cycles**, you can try use that so this PR goes out with the latest and greatest.

```suggestion
        value: quay.io/redhat-appstudio-tekton-catalog/task-buildah-remote-oci-ta:0.2@sha256:6fa1e7972e19c5bd8cc26c29c4de6ddb7f4f538dca10465bc4c1cb5a16400886
```

This requires some upgrade steps https://github.com/konflux-ci/build-definitions/blob/main/task/buildah-remote-oci-ta/0.2/MIGRATION.md

This will also require upgrading non-remote `buildah-oci-ta` also to 0.2, but the migration steps are the same. https://github.com/konflux-ci/build-definitions/blob/main/task/buildah-oci-ta/0.2/MIGRATION.md
Image reference for that: `quay.io/redhat-appstudio-tekton-catalog/task-buildah-oci-ta:0.2@sha256:812f2886915b2bd5ea306c179bb778703d47c95585a93953246439ffe3d2ca04`.

File: .tekton/basic-component-pipeline.yaml
Comment (msugakov) [inline .tekton/basic-component-pipeline.yaml:327]: If @Stringy found a way to make matrix work, it would be nice if you could apply that here to reduce duplication in ppc64le,s390x and arm64 tasks.

File: .tekton/basic-component-pipeline.yaml
Comment (msugakov) [inline .tekton/basic-component-pipeline.yaml:392]:
```suggestion
        values: [ "true" ]
```

File: .tekton/basic-component-pipeline.yaml
Comment (msugakov) [inline .tekton/basic-component-pipeline.yaml:419]:
```suggestion
        values: [ "true" ]
```

File: .tekton/basic-component-pipeline.yaml
Comment (msugakov) [inline .tekton/basic-component-pipeline.yaml:368]: Let's please apply `IMAGE_EXPIRES_AFTER` parameter on both `build-image-manifest`-s with the same value as for included images. We don't need manifests to stay in repos forever.

File: .tekton/basic-component-pipeline.yaml
Comment (msugakov) [inline .tekton/basic-component-pipeline.yaml:448]: These EC findings are interesting

image

See https://console.redhat.com/preview/application-pipeline/workspaces/rh-acs/applications/acs/pipelineruns/acs-pk7q4-vbvw2/security

Since giving it just `-amd64` wasn't sufficient, I suppose we'll need to make the full matrix - do the deprecated base image check for each arch.

File: .tekton/basic-component-pipeline.yaml
Comment (Molter73) [reply .tekton/basic-component-pipeline.yaml:327]: I've been following the conversations on this, my understanding is that he is rolling back matrices since they are not recommended ATM, so I'm skipping this comment.

File: .tekton/basic-component-pipeline.yaml
Comment (Molter73) [reply .tekton/basic-component-pipeline.yaml:448]: I think I have an alternative for this, we can feed the manifest URL and digest, the base image has the same digests for all archs, so it should scan check all archs (at least that's how I have it in #12055 and I think it worked but now I can't find the pipeline).

Comment (msugakov) [reply .tekton/basic-component-pipeline.yaml:448]: > the base image has the same digests for all archs

This is the base image manifest may have the same digest because it's just one manifest. As it's a manifest it will point to the actual images and those will have different digests.
Anyway, I'm interested in your approach and how would it work out.

**Comment by Molter73** [comment]
After some dicussion [here](https://redhat-internal.slack.com/archives/C079XCJEZK6/p1722263819380299), we've decided to rip operator-bundle into its own pipeline so that we can move forward onboarding the rest of components. I'll start working on a follow up for operator-bundle.

File: .tekton/basic-component-pipeline.yaml
Comment (Molter73) [reply .tekton/basic-component-pipeline.yaml:448]: Turns out when using the v0.2 buildah steps with the v0.4 version of this task, it can use the SBOM directly to scan images, so we can remove `BASE_IMAGE_DIGESTS` altogether.

File: .tekton/basic-component-pipeline.yaml
Comment (msugakov) [inline .tekton/basic-component-pipeline.yaml:207]: Now there's no reason for `basic-component-pipeline` to depend on or include the `output-image-tag-construction`. Could you please delete the entire `output-image-tag-construction` task from this pipeline and adjust result references from `tasks.output-image-tag-construction.results.FINAL_IMAGE_TAG` to `tasks.determine-image-tag.results.IMAGE_TAG`?

File: .tekton/basic-component-pipeline.yaml
Comment (msugakov) [inline .tekton/basic-component-pipeline.yaml:109]: [not about this line] Please adjust comment in the `init` task similar to https://github.com/stackrox/scanner/pull/1574/files#diff-d2c759f3ab7490911f1c024eed8612d5faa9e9a2b2ca8e26b063ab3c2791947cR125 (because there's no `apply-tags` thing any more).

Comment (msugakov) [reply .tekton/basic-component-pipeline.yaml:109]: This is still open. It's about the current line 120.

```yaml
# expected input. We later actually add this tag on a built image with apply-tags task.
```
should be
```yaml
# expected input. We later actually add this tag on a built image with build-image-manifest-konflux task.
```

File: .tekton/basic-component-pipeline.yaml
Comment (msugakov) [reply .tekton/basic-component-pipeline.yaml:207]: This one is also still open.

File: .tekton/basic-component-pipeline.yaml
Comment (msugakov) [inline .tekton/basic-component-pipeline.yaml:378]: Sorry, just spotted now.
Giles came up with a nicer alternative here
```suggestion
        - $(tasks.build-container-amd64.results.IMAGE_REF)
        - $(tasks.build-container-s390x.results.IMAGE_REF)
        - $(tasks.build-container-ppc64le.results.IMAGE_REF)
        - $(tasks.build-container-arm64.results.IMAGE_REF)
```

File: .tekton/basic-component-pipeline.yaml
Comment (msugakov) [inline .tekton/basic-component-pipeline.yaml:406]:
```suggestion
        - $(tasks.build-container-amd64.results.IMAGE_REF)
        - $(tasks.build-container-s390x.results.IMAGE_REF)
        - $(tasks.build-container-ppc64le.results.IMAGE_REF)
        - $(tasks.build-container-arm64.results.IMAGE_REF)
```

**Review by msugakov — APPROVED** [review]
lgtm

**Pull request merged**
